Add julia-repl-vterm recipe

shg opened the pull request:
### Brief summary of what the package does

Provides a major-mode for inferior Julia process that runs in vterm, and a minor-mode that extends julia-mode to support interaction with the inferior Julia process.

### Direct link to the package repository

https://github.com/shg/julia-repl-vterm.el

### Your association with the package

I am the author and the maintainer.

### Relevant communications with the upstream package maintainer

None needed.

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them


package-lint has complaints about inconsistency between the function names and the package prefix. Let me know if this needs to be addressed.

shg commented:
I have questions about naming of functions and package prefix.

There's a package named julia-repl.el already in MELPA. Is it ok to use the name julia-repl-vterm.el for mine? I think that the current name julia-repl-vterm.el is good because the name well explains what the package is. But if necessary, I will rename this package.

I use two function prefixes, inferior-julia-repl-vterm-* and julia-with-repl-vterm-*, both of which are different from the package name julia-repl-vterm. Package-lint complains of this. I will rename them if this must be fixed. But I am not sure what are the good prefixes. Just renaming all of them to julia-repl-vterm-* would make the roles of the functions obsucure. I'd appreciate if any of you have suggetsions, any comvensions, or pointer to other packages I can refer to.

riscy commented:
In this case I think it would be less ambiguous to use `julia-vterm-repl` as your package's name, just to clarify that they're separate packages.

> I use two function prefixes, inferior-julia-repl-vterm-* and julia-with-repl-vterm-*, both of which are different from the package name julia-repl-vterm. Package-lint complains of this. I will rename them if this must be fixed. But I am not sure what are the good prefixes. Just renaming all of them to julia-repl-vterm-* would make the roles of the functions obsucure. I'd appreciate if any of you have suggestions, any conventions, or pointer to other packages I can refer to.

Package-lint is right to complain -- it's our main tool for making sure that packages' functions and variables don't collide with each other too often.  I can see how you feel it would make the roles of the functions obscure, but it helps to think of the prefixes as a kind of "namespace".  Shortening your prefix might help (e.g.`julia-vterm`).

If the prefixes are troublesome, a few packages seem to exist that address the issue (although many aren't compatible with package-lint, so we're left to manually check the implied prefixes):

- https://melpa.org/#/nameless -- cosmetic only, package-lint stays happy
- https://melpa.org/#/with-namespace -- I've seen this one used recently
- https://melpa.org/#/names

shg commented:
Thank you @riscy .

I will change the package name to julia-vterm.el and use julia-vterm-* for the minor-mode.

I'm still thinking what the name for the major-mode for inferior process should be. Using inferior-*-mode as the name of the major mode for repl is a sort of the standard in existing packages. Package-lint complain about it but some packages that use such names/prefixes have been accepted in MELPA. Also run-<langname> is the commonly used function name to run an inferior process, and it also does not conform to package-lint criteria.

I like to know whether we should always follow package-lint's criteria, or in some cases following conventions is preferred. I am not criticizing package-lint. I just want to know the criteria because there seem some exceptions.

I may use julia-vterm-repl-* (package-lint conformable) or inferior-julia-vterm-* (maybe more conventional) accordingly.

I appreciate your comments and suggestions.

riscy commented:
> Using inferior-*-mode as the name of the major mode for repl is a sort of the standard in existing packages. Package-lint complain about it but some packages that use such names/prefixes have been accepted in MELPA. Also run-langname is the commonly used function name to run an inferior process, and it also does not conform to package-lint criteria.

Yeah, package-lint will let you get away with some Emacs idioms.  Allowing for, e.g., `global-<package-name>-mode` is an example of a convention that package-lint has adopted.  This list has evolved over time, and the typical MELPA review only considers the very latest version of package-lint.  Rare exceptions are sometimes made for large legacy packages.

I'd expect there are many packages on MELPA that violate present-day guidelines.  Ideally the maintainers will try to fix those as they move forward, and to keep on top of the latest version of package-lint -- or at least not make things worse. :)

shg commented:
Thank you again @riscy . I updated the file and now package-lint does not report any issue.

Package name is now julia-vterm.el, and I changed the prefix of the major-mode to julia-vterm-* and that of the minor-mode to julia-vterm-repl-*.

Direct link to the package repository

https://github.com/shg/julia-vterm.el

riscy commented:
Thanks, this looks good.  The last thing I'd note is that you're relying on a private variable from vterm.el, `vterm--process`.  It might be a good idea to reach out to the maintainers and raise an issue about making that interface public so that you (and others) are taking less of a risk by depending on it.

shg commented:
Thank you very much @riscy . I will talk to emacs-libvterm maintainers about the issue you pointed out.
